src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt

```diff
@@ -220,11 +220,13 @@ fun processPathResolveVariants(lookup: ImplLookup, path: RsPath, isCompletion: B
     val crateRoot = path.crateRoot
     if (!path.hasColonColon) {
         if (Namespace.Types in ns) {
-            if (processor("self", containingMod)) return true
+            val context = parent?.contextOfType<RsValueParameter>()
+            if (context is RsValueParameter) {
+                if (processor("self", context)) return true
+            } else if (processor("self", containingMod)) return true
+
             val superMod = containingMod.`super`
-            if (superMod != null) {
-                if (processor("super", superMod)) return true
-            }
+            if (superMod != null && processor("super", superMod)) return true
```

Comment:
It still seems to me that it is wrong to modify name resolution hear, because semantically resolving a self parameter to itself does not make sense.

I think that modifying insertion handler to handle self parameters simlarly to use globs should be enough:


```Rust
private fun doubleColonLookup(base: LookupElementBuilder): LookupElementBuilder =
    base.withTailText("::")
        .withInsertHandler({ ctx, _ ->
            val offset = ctx.tailOffset - 1
            val inUse = PsiTreeUtil.findElementOfClassAtOffset(ctx.file, offset, RsUseGlobList::class.java, false) != null
            val inSelfParam = PsiTreeUtil.findElementOfClassAtOffset(ctx.file, offset, RsSelfParameter::class.java, false) != null
            if (!(inUse || inSelfParam)) {
                ctx.addSuffix("::")
            }
        })
```

Reply:
For me, it was unclear, how lookup builder work.